distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,856 @@
+        offset = self._offset
+        buffers = self._buffers
+        while size:
+            b = buffers[0]
+            b_len = len(b) - offset
+            if b_len <= size:
+                buffers.popleft()
+                size -= b_len
+                offset = 0
+            else:
+                offset += size
+                break
+        self._offset = offset
+
+    def write(self, data):
+        transport = self.transport
+
+        if transport._eof:
+            raise RuntimeError("Cannot call write() after write_eof()")
+        if transport._empty_waiter is not None:
+            raise RuntimeError("unable to write; sendfile is in progress")
+        if not data:
+            return
+        if transport._conn_lost:
+            return
+
+        if not self._buffers:
+            try:
+                n = transport._sock.send(data)
+            except (BlockingIOError, InterruptedError):
+                pass
+            except (SystemExit, KeyboardInterrupt):
+                raise
+            except BaseException as exc:
+                transport._fatal_error(exc, "Fatal write error on socket transport")
+                return
+            else:
+                data = data[n:]
+                if not data:
+                    return
+            # Not all was written; register write handler.
+            transport._loop._add_writer(transport._sock_fd, self._on_write_ready)
+
+        # Add it to the buffer.
+        self._buffer_append(data)
+        transport._maybe_pause_protocol()
+
+    def writelines(self, buffers):
+        waiting = bool(self._buffers)
+        for b in buffers:
+            self._buffer_append(b)
+        if not waiting:
+            try:
+                self._do_bulk_write()
+            except (BlockingIOError, InterruptedError):
+                pass
+            except (SystemExit, KeyboardInterrupt):
+                raise
+            except BaseException as exc:
+                self.transport._fatal_error(
+                    exc, "Fatal write error on socket transport"
+                )
+                return
+            if not self._buffers:
+                return
+            # Not all was written; register write handler.
+            self.transport._loop._add_writer(
+                self.transport._sock_fd, self._on_write_ready
+            )
+
+        self.transport._maybe_pause_protocol()
+
+    def is_closing(self):
+        return self.transport.is_closing()
+
+    def close(self):
+        return self.transport.close()
+
+    def abort(self):
+        return self.transport.abort()
+
+    def get_extra_info(self, key):
+        return self.transport.get_extra_info(key)
+
+    def _do_bulk_write(self):
+        buffers = self._buffer_peek()
+        if len(buffers) == 1:
+            n = self.transport._sock.send(buffers[0])
+            self._buffer_advance(n)
+        else:
+            n = self.transport._sock.sendmsg(buffers)
+            self._buffer_advance(n)
```

Comment:
From https://stackoverflow.com/a/4259888/17100540 it sounds like `send` might just be `sendmsg` internally. I have no idea for real in cpython, just wondering if this distinction for the 1-length case is necessary.

Reply:
cpython does some extra work for sendmsg, and not all platforms support sendmsg. Seems fine to use both where applicable.